pkg/identity/cache/local.go

```diff
@@ -0,0 +1,189 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"fmt"
+
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/kvstore"
+	"github.com/cilium/cilium/pkg/kvstore/allocator"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+type localIdentityCache struct {
+	mutex               lock.RWMutex
+	identitiesByID      map[identity.NumericIdentity]*identity.Identity
+	identitiesByLabels  map[string]*identity.Identity
+	nextNumericIdentity identity.NumericIdentity
+	minID               identity.NumericIdentity
+	maxID               identity.NumericIdentity
+	events              allocator.AllocatorEventChan
+}
+
+func newLocalIdentityCache(minID, maxID identity.NumericIdentity, events allocator.AllocatorEventChan) *localIdentityCache {
+	return &localIdentityCache{
+		identitiesByID:      map[identity.NumericIdentity]*identity.Identity{},
+		identitiesByLabels:  map[string]*identity.Identity{},
+		nextNumericIdentity: minID,
+		minID:               minID,
+		maxID:               maxID,
+		events:              events,
+	}
+}
+
+func (l *localIdentityCache) bumpNextNumericIdentity() {
+	if l.nextNumericIdentity == l.maxID {
+		l.nextNumericIdentity = l.minID
+	} else {
+		l.nextNumericIdentity++
+	}
+}
+
+// allocateNumericIdentity returns the next available numeric identity or an error
+func (l *localIdentityCache) allocateNumericIdentity() (identity.NumericIdentity, error) {
+	firstID := l.nextNumericIdentity
+	for {
+		if _, ok := l.identitiesByID[l.nextNumericIdentity]; !ok {
+			id := l.nextNumericIdentity
+			l.bumpNextNumericIdentity()
+			return id, nil
+		}
+
+		l.bumpNextNumericIdentity()
+		if l.nextNumericIdentity == firstID {
+			return 0, fmt.Errorf("out of local identity space")
+		}
+	}
+}
+
+// lookupOrCreate searches for the existence of a local identity with the given
+// labels. If it exists, the reference count is incremented and the identity is
+// returned. If it does not exist, a new identity is created with a unique
+// numeric identity. All identities returned by lookupOrCreate() must be
+// released again via localIdentityCache.release().
+func (l *localIdentityCache) lookupOrCreate(lbls labels.Labels) (*identity.Identity, bool, error) {
+	l.mutex.Lock()
+	defer l.mutex.Unlock()
+
+	stringRepresentation := string(lbls.SortedList())
+	if id, ok := l.identitiesByLabels[stringRepresentation]; ok {
+		id.ReferenceCount++
+		return id, false, nil
+	}
+
+	numericIdentity, err := l.allocateNumericIdentity()
+	if err != nil {
+		return nil, false, err
+	}
+
+	id := &identity.Identity{
+		ID:             numericIdentity | identity.LocalIdentityFlag,
+		Labels:         lbls,
+		LabelArray:     lbls.LabelArray(),
+		ReferenceCount: 1,
+	}
+
+	l.identitiesByLabels[stringRepresentation] = id
+	l.identitiesByID[numericIdentity] = id
```

Comment:
IMO it would be clearer and less error-prone to have the map key include the local identity flag. Maybe it should already be returned by the "allocator"?